File: benchmarks/generic_collections/hash_table.cpp
#include <cstdint>
#include <cstdio>
#include <vector>

static std::int32_t lookup_or(const std::vector<std::int32_t>& keys,
                              const std::vector<std::int32_t>& values,
                              std::int32_t key,
                              std::int32_t fallback) {
  for (std::size_t i = 0; i < keys.size() && i < values.size(); ++i) {
    if (keys[i] == key) {
      return values[i];
    }
  }
  return fallback;
}

int main() {
  constexpr std::int32_t n = 1024;
  std::vector<std::int32_t> keys(n);
  std::vector<std::int32_t> values(n);
  for (std::int32_t i = 0; i < n; ++i) {
    keys[static_cast<std::size_t>(i)] = i * 2 + 1;
    values[static_cast<std::size_t>(i)] = i + 7;
  }

  std::int32_t checksum = 0;
  for (std::int32_t q = 0; q < n; ++q) {
    checksum += lookup_or(keys, values, q * 2 + 1, 0);
  }
  std::printf("%d\n", checksum);
  return checksum == 530944 ? 0 : 1;
}

File: benchmarks/generic_collections/hash_table.rs
fn lookup_or(keys: &[i32], values: &[i32], key: i32, fallback: i32) -> i32 {
    let limit = keys.len().min(values.len());
    for i in 0..limit {
        if keys[i] == key {
            return values[i];
        }
    }
    fallback
}

fn main() {
    let n: i32 = 1024;
    let mut keys = vec![0_i32; n as usize];
    let mut values = vec![0_i32; n as usize];
    for i in 0..n {
        keys[i as usize] = i * 2 + 1;
        values[i as usize] = i + 7;
    }

    let mut checksum = 0_i32;
    for q in 0..n {
        checksum += lookup_or(&keys, &values, q * 2 + 1, 0);
    }
    println!("{}", checksum);
    std::process::exit(if checksum == 530944 { 0 } else { 1 });
}
